pythonPackages.diagrams: init at 0.17.0

addict3d opened the pull request:
###### Motivation for this change
Add python diagrams library. Generate nice architecture diagrams
with python code, allowing for rapid iteration.

The images are rendered using dot from graphviz. It outputs your choice of `[png jpg svg pdf]`

Closes #101532

###### Things done
I generated some diagrams using this library, with both of the following shell expressions. There are [many examples](https://diagrams.mingrammer.com/docs/getting-started/examples) on the project's website.

```
nix-shell -I nixpkgs=./ -p python3 python3Packages.diagrams python3Packages.graphviz
```
```
nix-shell -I nixpkgs=./ -E 'with import <nixpkgs> {}; (python3.withPackages (ps: [ps.diagrams ps.graphviz])).env'
```

The closure size is ~150M.
The library itself has ~20M in images.

evanjs commented:
Please move the `maintainers.list` change into a separate commit

evanjs commented on pkgs/development/python-modules/diagrams/default.nix:
```diff
@@ -0,0 +1,69 @@
+{ lib
+, fetchFromGitHub
+, substituteAll
```
This can be removed

evanjs commented:
Perhaps this could be added to `all-packages` as well.

e.g.  https://github.com/NixOS/nixpkgs/blob/91358d4c50d802734bb076d712cfde52057c8233/pkgs/top-level/all-packages.nix#L22650

addict3d committed:
pythonPackages.diagrams: init at 0.17.0

Add python diagrams library. Generate nice architecture diagrams
with python code, allowing for rapid iteration.

The images are rendered using dot from graphviz.

Closes #101532

addict3d committed:
maintainers: add addict3d

addict3d commented:
> Perhaps this could be added to `all-packages` as well.
> 
> e.g.
> 
> https://github.com/NixOS/nixpkgs/blob/91358d4c50d802734bb076d712cfde52057c8233/pkgs/top-level/all-packages.nix#L22650

Oh cool. I wasn't aware of that. Added it.

Rebased to resolve the suggestions. I've tried my best to follow the nixpkgs and python-modules conventions and am happy to learn more and make any changes.

mweinelt commented on pkgs/development/python-modules/diagrams/default.nix:
```diff
@@ -0,0 +1,68 @@
+  version = "0.17.0";
+  disabled = ! isPy3k;
+
+  # TODO consider packaging from Github: https://github.com/mingrammer/diagrams
```
This should be easy. Use `fetchFromGitHub` instead of `fetchPypi` and set `format = "pyproject";`. I assume we would then have the `tests/` directory.

addict3d replied:
I spent a few hours working on this. I am far from a python expert, but it seems like the GitHub repo is structurally quite different from the pypi tarball. I will upload a new branch in a bit with my (unsuccessful) efforts.

addict3d replied:
You can see my (stalled) progress on packaging from GitHub here: https://github.com/NixOS/nixpkgs/pull/102060

@evanjs also had what is probably a terrible idea for tests, which is to 'cherry-pick' just the tests from GitHub...

mweinelt reviewed (changes requested):
That looks pretty good already. 

It is customary for python packages to get the `checkPhase` running. Currently no tests are being found, even though upstream has tests. This is because the PyPi Package doesn't include any.

```
----------------------------------------------------------------------
Ran 0 tests in 0.001s
```

Also the default `checkPhase` might not work for this package. To find out how the test suite is run it is usually helpful to check how their CI is set up.

https://github.com/mingrammer/diagrams/blob/master/.github/workflows/test.yml#L35

Please try and get the tests running.

addict3d commented:
Thanks. I will take a look at it.

On Sun, Oct 25, 2020, 11:13 AM Martin Weinelt <notifications@github.com>
wrote:

> *@mweinelt* requested changes on this pull request.
>
> That looks pretty good already.
>
> It is customary for python packages to get the checkPhase running.
> Currently no tests are being found, even though upstream has tests. This is
> because the PyPi Package doesn't include any.
>
> ----------------------------------------------------------------------
> Ran 0 tests in 0.001s
>
> Also the default checkPhase might not work for this package. To find out
> how the test suite is run it is usually helpful to check how their CI is
> set up.
>
>
> https://github.com/mingrammer/diagrams/blob/master/.github/workflows/test.yml#L35
>
> Please try and get the tests running.
> ------------------------------
>
> In pkgs/development/python-modules/diagrams/default.nix
> <https://github.com/NixOS/nixpkgs/pull/101533#discussion_r511609050>:
>
> > @@ -0,0 +1,68 @@
> +{ lib
> +, fetchFromGitHub
> +, buildPythonPackage
> +, fetchPypi
> +, isPy3k
> +, graphviz
> +}:
> +
> +buildPythonPackage rec {
> +  pname = "diagrams";
> +  version = "0.17.0";
> +  disabled = ! isPy3k;
> +
> +  # TODO consider packaging from Github: https://github.com/mingrammer/diagrams
>
> This should be easy. Use fetchFromGitHub instead of fetchPypi and set format
> = "pyproject";. I assume we would then have the tests/ directory.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/101533#pullrequestreview-516368396>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAMAS4NKIHC3VS6HL72ADLSMQ6A3ANCNFSM4S5MGCDQ>
> .
>

sikmir commented on pkgs/top-level/all-packages.nix:
```diff
@@ -3145,6 +3145,8 @@ in
 
   dev86 = callPackage ../development/compilers/dev86 { };
 
+  diagrams = with python3.pkgs; toPythonApplication diagrams;
```
@jonringer I just wonder, is it good practice to duplicate python packages listed in python-packages.nix that have some cli tools to all-packages.nix?

evanjs replied:
In the case of OpenRazer, I think it was appropriate due to the nature of the package, but I am also curious about best practices, especially in this case.

mweinelt replied:
This makes it so you can install `diagrams` and have the executable, as well as use `python3Packages.diagrams` if you want to import the library.

Read up on it here https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#topythonapplication-function.

addict3d replied:
I reverted this because, for this diagrams library, it does not make sense as an executable.

addict3d commented:
For me personally, I'd make use of this through a python environment,
rather than using it from top-level.

Happy to change the PR to match best practices.


On Tue, Oct 27, 2020, 1:16 PM Evan Stoll <notifications@github.com> wrote:

> *@evanjs* commented on this pull request.
> ------------------------------
>
> In pkgs/top-level/all-packages.nix
> <https://github.com/NixOS/nixpkgs/pull/101533#discussion_r512877362>:
>
> > @@ -3145,6 +3145,8 @@ in
>
>    dev86 = callPackage ../development/compilers/dev86 { };
>
> +  diagrams = with python3.pkgs; toPythonApplication diagrams;
>
> In the case of OpenRazer, I think it was appropriate due to the nature of
> the package, but I am also curious about best practices, especially in this
> case.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/101533#discussion_r512877362>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAMASZF6U77QIWDNTOH2BTSM355TANCNFSM4S5MGCDQ>
> .
>

addict3d commented:
Here is my attempt to package this from the upstream GitHub instead of the pypi tarball. Currently this produces an empty python module. I'm not (yet) familiar with python packaging to understand what is happening.

https://github.com/addict3d/nixpkgs/tree/add-python-diagrams-from-github

It appears that the library code is generated correctly and the tests all pass, but getting the library working is important.

addict3d commented:
I removed the top-level/all-packages entry because it doesn't make sense as a binary.

You need a python environment with this library, and then to produce an image, call `python my-diagram-code.py` (where your code imports diagrams).

evanjs commented:
Since the `setup.py` and `tests` directories don't seem to both be available from a single source, I was wondering if it's _acceptable_ to pull in both sources.

Not quite sure what else can be done if we can't determine how the GitHub sources can be installed or etc.
Anyway, the following seems to work, but feels awfully hacky to me.

I also wonder if sourcing GitHub and pulling the `setup.py` from Pypi would be any better 🤷‍♂️ 

<details>
<summary>sources from Pypi with tests from GitHub</summary>

``` diff
diff --git a/pkgs/development/python-modules/diagrams/default.nix b/pkgs/development/python-modules/diagrams/default.nix
index 1f05abab32c..bbfa50a5a3c 100644
--- a/pkgs/development/python-modules/diagrams/default.nix
+++ b/pkgs/development/python-modules/diagrams/default.nix
@@ -4,17 +4,28 @@
 , fetchPypi
 , isPy3k
 , graphviz
+, pytestCheckHook
 }:

-buildPythonPackage rec {
+let
   pname = "diagrams";
   version = "0.17.0";
+  gitHubSrc = fetchFromGitHub {
+    owner = "mingrammer";
+    repo = pname;
+    rev = "v${version}";
+    sha256 = "1h5i9xb4fsmy81sznvqycxl712d9k695vznkpx2wm1ki7hvz86s8";
+  };
+in
+buildPythonPackage rec {
   disabled = ! isPy3k;
+  inherit pname version;

   # TODO consider packaging from Github: https://github.com/mingrammer/diagrams
   # note: outscale package is broken, __init__.py is missing
   src = fetchPypi {
     inherit pname version;
+
     sha256 = "17v5nzswva74bawxrvkza8a5d9aqyhb2gkncp3fqv6rszmk62hw6";
   };

@@ -22,6 +33,8 @@ buildPythonPackage rec {
     ./diags_fix_aliases.patch
   ];

+  checkInputs = [ pytestCheckHook ];
+
   # This script is patching setup.py from the pypi tarball, to retain
   # the png images which this library uses to generate diagrams.
   postPatch = ''
@@ -55,6 +68,8 @@ buildPythonPackage rec {
       done
       echo "    ],)"
     } >> setup.py
+
+    cp -r ${gitHubSrc}/tests .
   '';

   buildInputs = [ graphviz ];
```

</details>

mweinelt commented:
I thought we could fetch this from GitHub and set `format = "pyproject";` so it would build from the `pyproject.toml`. We don't strictly need a setup.py. But IIRC the following part is missing in `pyproject.toml`.

```
[build-system]
requires = ["poetry-core>=1.0.0"]
build-backend = "poetry.core.masonry.api"
```

The author only uses this with poetry to develop, build and publish, so they don't need that explicitly.

The test phase uses the unittest module: https://github.com/mingrammer/diagrams/blob/master/.github/workflows/test.yml#L35

mweinelt commented:
Builds after patching the upstream git checkout like this:

```
diff --git a/pyproject.toml b/pyproject.toml
index 8725f84..c134ab8 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -11,8 +11,8 @@ include = ["resources/**/*"]
 
 [tool.poetry.dependencies]
 python = "^3.6"
-graphviz = "^0.13.2"
-jinja2 = "^2.10"
+graphviz = "^0.14.1"
+jinja2 = "^2.11"
 contextvars = { version = "^2.4", python = "~3.6" }
 
 [tool.poetry.dev-dependencies]
@@ -24,3 +24,7 @@ isort = "^4.3"
 
 [tool.black]
 line-length = 120
+
+[build-system]
+requires = ["poetry_core>=1.0.0"]
+build-backend = "poetry.core.masonry.api"
```

martinbaillie commented:
hey @addict3d are you still planning on taking this package into Nixpkgs? I've been using your branch for a while and would be keen to see it merged

SuperSandro2000 commented on pkgs/development/python-modules/diagrams/default.nix:
```diff
@@ -0,0 +1,68 @@
+buildPythonPackage rec {
+  pname = "diagrams";
+  version = "0.17.0";
+  disabled = ! isPy3k;
```
```suggestion
  disabled = !isPy3k;
```

addict3d replied:
Thank you for looking at this PR SuperSandro2000 😀

addict3d replied:
In the GH branch, I switched this to `pythonOlder` [link, see L#19](https://github.com/NixOS/nixpkgs/pull/102060/files#diff-dcdb3f0a79fc374de9f7df1479a0c343bb35c7e8c2903ac35e2e336c18b1001e)

SuperSandro2000 commented on pkgs/development/python-modules/diagrams/default.nix:
```diff
@@ -0,0 +1,68 @@
+      --replace ", 'jinja2>=2.10,<3.0'" ""  # not used when building from the source tarball
+
+    # scan dirs for pngs, building up the (directory, files) structure for data_files
+    {
```
Please convert this into a patch.

addict3d replied:
this code is no longer necessary in the other branch

addict3d commented:
> hey @addict3d are you still planning on taking this package into Nixpkgs? I've been using your branch for a while and would be keen to see it merged

oh. err sorry about that. yes I will get back to this and address the comments

addict3d commented:
I'm closing this in favor of packaging directly from upstream (GitHub) rather than pypi. This way the package will include tests, and it removes one layer of indirection.

See https://github.com/NixOS/nixpkgs/pull/102060
